File: agent-sdk/typescript/examples/01-echo.ts
// Minimal echo agent — replies to every prompt with `echo: <prompt text>`.
//
// The shortest runnable demonstration of the `AgentService` host API.
// Use this as a smoke target while iterating on a caller, or as a
// starting shape when writing your own agent.
//
// Shared connection resolution (`_connection.ts`): $NATS_CONTEXT wins as one
// complete source; otherwise $NATS_URL combines with NATS_NKEY_SEED_FILE or
// NATS_CREDS, then falls back to localhost. NATS_SENDER_IDENTITY=signed derives
// registration identity from that same source; identity is off by default.

import { AgentService } from "@synadia-ai/agent-service";
import { openExampleNatsConnection, waitForTermination } from "./_connection";

async function main(): Promise<void> {
  const connection = await openExampleNatsConnection();
  let service: AgentService | undefined;

  try {
    // Identity → subject `agents.prompt.echo.<owner>.<name>`. Owner and name are
    // env-overridable so several people can run this against one server without
    // colliding; `agent` ("echo") is what this example *is*, so it stays fixed.
    const heartbeatIntervalS = Number(process.env["NATS_AGENT_HEARTBEAT_INTERVAL"]) || undefined;
    service = new AgentService({
      nc: connection.nc,
      agent: "echo",
      owner:
        process.env["SYNADIA_ECHO_OWNER"] ??
        process.env["SYNADIA_OWNER"] ??
        process.env["NATS_AGENT_OWNER"] ??
        process.env["USER"] ??
        "anon",
      name:
        process.env["SYNADIA_ECHO_NAME"] ??
        process.env["SYNADIA_NAME"] ??
        process.env["NATS_AGENT_NAME"] ??
        "main",
      ...(heartbeatIntervalS !== undefined ? { heartbeatIntervalS } : {}),
      ...(connection.signer ? { identity: { signer: connection.signer } } : {}),
      description: "Echo agent — replies with the prompt prefixed by 'echo: '",
    });

    service.onPrompt(async (envelope, response) => {
      await response.send(`echo: ${envelope.prompt}`);
    });

    await service.start();
    console.log(`echo agent listening on ${service.subject.prompt}`);
    console.log("press Ctrl+C to stop");
    await waitForTermination();
    console.log("\nshutting down…");
  } finally {
    try {
      await service?.stop();
    } finally {
      await connection.close();
    }
  }
}

void main().catch((err: unknown) => {
  console.error("echo agent failed:", err);
  process.exit(1);
});

File: agent-sdk/python/examples/01-echo.py
# 01 · Echo agent — the smallest Synadia Agent Protocol agent (Python).
#
# Replies to every prompt with `echo: <prompt text>`. No LLM, no state — just
# enough to show the shape of an agent built on AgentService: connect to NATS,
# construct the service, handle prompts, start, shut down. Python mirror of
# agent-sdk/typescript/examples/01-echo.ts.
#
# Identity → subject agents.prompt.echo.<owner>.<session_name>. Owner and the
# session name are overridable (--owner / --session-name, or $NATS_AGENT_OWNER /
# $NATS_AGENT_NAME) so several people can run this against one server without
# colliding. --heartbeat-interval / $NATS_AGENT_HEARTBEAT_INTERVAL tunes the
# heartbeat cadence (default 30s).
#
# Connection: --context / --url, else $NATS_CONTEXT / $NATS_URL, else the
# selected `nats` context. Run with -h for the full flag list.

from __future__ import annotations

import argparse
import asyncio
import signal
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from synadia_ai.agents import Envelope

from examples._connect_cli import (
    add_agent_identity_flags,
    add_connection_flags,
    connect_from_cli,
)
from synadia_ai.agent_service import AgentService, PromptStream


async def main() -> None:
    parser = argparse.ArgumentParser(
        description="Echo agent — replies with the prompt prefixed by 'echo: '."
    )
    add_connection_flags(parser)
    add_agent_identity_flags(parser)
    args = parser.parse_args()

    nc = await connect_from_cli(args)

    service = AgentService(
        agent="echo",
        owner=args.owner,
        session_name=args.session_name,
        nc=nc,
        description="Echo agent — replies with the prompt prefixed by 'echo: '",
        heartbeat_interval_s=args.heartbeat_interval,
    )

    # The whole agent. Every incoming prompt runs this handler; whatever we
    # stream.send(...) is the reply. Here we send one chunk; an LLM agent would
    # send many, token by token.
    async def handler(envelope: Envelope, stream: PromptStream) -> None:
        await stream.send(f"echo: {envelope.prompt}")

    service.on_prompt(handler)
    await service.start()
    print(f"echo agent listening on {service.subject.prompt}")
    print("press Ctrl+C to stop")

    # add_signal_handler is the asyncio-safe way to wake `await stop.wait()`;
    # signal.signal + Event.set() does not reliably notify the running loop.
    loop = asyncio.get_running_loop()
    stop = asyncio.Event()
    for _sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(_sig, stop.set)
    try:
        await stop.wait()
    finally:
        print("\nshutting down…")
        await service.stop()
        await nc.close()


if __name__ == "__main__":
    asyncio.run(main())
